### ISSUE TITLE
Add tutorial on LightGBM with early stopping in probatus

### ISSUE THREAD

**Issue by Matgrb** [ing-bank/probatus#39]
- Add tutorial on LightGBM with early stopping in probatus

**Comment by operte** [comment]
I have a related question: Can you even use early stopping with `ShapRFECV` and a `LGBMClassifier`? `LGBMClassifier` only takes the `early_stopping_rounds` argument in the `.fit` method. The `ShapRFECV.fit` method does not accept arguments to pass to the `clf.fit` method. Am I overlooking something or is there another way to implement early stopping?

**Comment by Matgrb** [comment]
Yes you can, however, you need to write a class that inherits from `LGBMClassifier` and overwrites it's `fit` method and has the `early_stopping_rounds` as instance parameter, which you can optimize using `GridSearchCV` (.set_params() method should be able to overwrite it)

It is a bit of a hack, but it is not much of code required. If you are interested in implementing it, I am happy to help you with it.

**Comment by operte** [comment]
Good one, I might try this later and then I'll ask for some help :)

Is it uncommon to require passing parameters to the model's `.fit` method? If not, would a better solution be to change the `ShapRFECV.fit` method to allow passing arguments to the model's `.fit`?

**Comment by Matgrb** [comment]
The issue is that we want to use RandomSearchCV and GridSearchCV, and they have a specific way of calling the model.

For each parameter combination they .set_params() and then run cross_val_score(). If we want to optimize amount of early_stopping as well we need to have it as a parameter, or at least we need to make sure that cross_val_score() works correctly with it. 

So far that is the easiest way i found to overcomming it, but i am also interested if you have some other ideas. What i really would like, is to make sure that we don't put too much extra logic to ShapRFECV just to support it, while every model e.g. XGBClassifier and LGBMClassifier might update their API, causing errors in probatus in the future and extra maintainance work. I think just giving users example how to do it for LGBMClassifier should be sufficient.

**Comment by operte** [comment]
It seems that both [`GridSearchCV`](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html#sklearn.model_selection.GridSearchCV.fit) and [`RandomSearchCV`](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RandomizedSearchCV.html#sklearn.model_selection.RandomizedSearchCV.fit) both accept additional arguments to be passed to the model's `fit` method. So in principle this should work. 

Another issue I encountered, though, is that to use `early_stopping_rounds` one must also pass a `eval_set` to `LGBMClassifier`. `eval_set` will be different for each CV round, so the CV classes should also be able to pass the train and validation sets through this argument. Given this, do you still think that creating a child class of `LGBMClassifier` would work? It is not obvious to me how this could be implemented to make it work with the sklearn CV classes.

One alternative is to define a separate test set just for the `LGBMClassifier` to do its checks, like [here](https://github.com/Microsoft/LightGBM/issues/1044#issuecomment-546089433). However this is not very elegant and it requires splitting the data yet again, which we don't always have the luxury to do. 

Another idea is to pass the native LightGBM CV class to `ShapRFECV`, like suggested [here](https://github.com/Microsoft/LightGBM/issues/1044#issuecomment-343391649). I haven't checked this option, but from what I understood from the documentation, `ShapRFECV` does accept custom CV classes, doesn't it?

**Comment by Matgrb** [comment]
That is indeed an issue, especially if you have a specific grouping of samples that you want to include in your cross-validation. But i would say that simple case of just using randomly selected 10/20% of train folds as validation should be sufficient for most of the use cases. This can be done quite easily by overvriting the `fit()` method of LGBM, and leaving some part of `X_train` and `y_train` as eval_set, before calling the `super().fit()`. I think this would be the most elegant way, and at least would give users hint on what to do. They can apply more complex methods if they want on top of the example we give.

**Comment by operte** [comment]
I had to do this for a project. I decided to overload ShapRFECV instead of overloading LGBM. The modifications were quite simple: 

```
class ShapRFECVES(ShapRFECV):
    """Recursive Feature Elimination with Shapley values.
    
    This is a child of ShapRFECV, defined in the package
    [Probatus](https://github.com/ing-bank/probatus/blob/main/probatus/feature_elimination/feature_elimination.py)
  
    We have made it compatible with LightGBM's early stopping functionality, to prevent major overfitting on the train
    set.
  
    Parent docstring:
    """
    __doc__ = __doc__ + ShapRFECV.__doc__
  
    def __init__(self, clf, early_stopping_rounds=30, **kwargs):
        """This method initializes the class.

        Args:
            clf (binary classifier, GridSearchCV or RandomizedSearchCV):
                A model that will be optimized and trained at each round of features elimination. The recommended model
                is [LGBMClassifier](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.LGBMClassifier.html),
                because it by default handles the missing values and categorical variables. This parameter also supports
                [GridSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html)
                and [RandomizedSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RandomizedSearchCV.html).
            step (int or float, optional):
                Number of lowest importance features removed each round. If it is an int, then each round such number of
                features is discarded. If float, such percentage of remaining features (rounded down) is removed each
                iteration. It is recommended to use float, since it is faster for a large number of features, and slows
                down and becomes more precise towards less features. Note: the last round may remove fewer features in
                order to reach min_features_to_select.
                If columns_to_keep parameter is specified in the fit method, step is the number of features to remove after 
                keeping those columns.
            min_features_to_select (int, optional):
                Minimum number of features to be kept. This is a stopping criterion of the feature elimination. By
                default the process stops when one feature is left. If columns_to_keep is specified in the fit method,
                it may overide this parameter to the maximum between length of columns_to_keep the two.
            cv (int, cross-validation generator or an iterable, optional):
                Determines the cross-validation splitting strategy. Compatible with sklearn
                [cv parameter](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFECV.html).
                If None, then cv of 5 is used.
            scoring (string or probatus.utils.Scorer, optional):
                Metric for which the model performance is calculated. It can be either a metric name  aligned with
                predefined [classification scorers names in sklearn](https://scikit-learn.org/stable/modules/model_evaluation.html).
                Another option is using probatus.utils.Scorer to define a custom metric.
            n_jobs (int, optional):
                Number of cores to run in parallel while fitting across folds. None means 1 unless in a
                `joblib.parallel_backend` context. -1 means using all processors.
            verbose (int, optional):
                Controls verbosity of the output:
                - 0 - nether prints nor warnings are shown
                - 1 - 50 - only most important warnings
                - 51 - 100 - shows other warnings and prints
                - above 100 - presents all prints and all warnings (including SHAP warnings).
            random_state (int, optional):
                Random state set at each round of feature elimination. If it is None, the results will not be
                reproducible and in random search at each iteration a different hyperparameters might be tested. For
                reproducible results set it to integer.
            early_stopping_rounds (int):
                The number of rounds to check improvement. If no improvement after this many rounds, we stop training
        """
        super(ShapRFECVES, self).__init__(clf, **kwargs)
        self.early_stopping_rounds = early_stopping_rounds
  
    def _get_feature_shap_values_per_fold(self, X, y, clf, train_index, val_index, scorer, verbose=0):
        """Calculate the shap values on validation set, and Train and Val score.
  
        Args:
            X (pd.DataFrame):
                Dataset used in CV.
            y (pd.Series):
                Binary labels for X.
            clf (binary classifier):
                Model to be fitted on the train folds.
            train_index (np.array):
                Positions of train folds samples.
            val_index (np.array):
                Positions of validation fold samples.
            scorer (string, callable or None):
                A string (see sklearn [model scoring](https://scikit-learn.org/stable/modules/model_evaluation.html))
                or
                a scorer callable object, function with the signature `scorer(estimator, X, y)`.
            verbose (int, optional):
                Controls verbosity of the output:
                - 0 - neither prints nor warnings are shown
                - 1 - 50 - only most important warnings regarding data properties are shown (excluding SHAP warnings)
                - 51 - 100 - shows most important warnings, prints of the feature removal process
                - above 100 - presents all prints and all warnings (including SHAP warnings).
  
        Returns:
            (np.array, float, float):
                Tuple with the results: Shap Values on validation fold, train score, validation score.
        """
        X_train, X_val = X.iloc[train_index, :], X.iloc[val_index, :]
        y_train, y_val = y.iloc[train_index], y.iloc[val_index]
  
        # Fit model with train folds
        clf = clf.fit(
            X_train,
            y_train,
            eval_set=[(X_train, y_train), (X_val, y_val)],
            eval_metric="auc",
            verbose=2000,
            early_stopping_rounds=self.early_stopping_rounds,
        )
  
        # Score the model
        score_train = scorer(clf, X_train, y_train)
        score_val = scorer(clf, X_val, y_val)
  
        # Compute SHAP values
        shap_values = shap_calc(clf, X_val, verbose=verbose)
        return shap_values, score_train, score_val`
```

What do you think of this approach? Only the `_get_feature_shap_values_per_fold` method had to be modified.

**Comment by Matgrb** [comment]
I really like the idea, i haven't thought about this.

In that case we have two options:
- Modify ShapRFECV to have additional parameter `early_stopping_rounds`. The drawback of this method is that the parameter will only work correctly with Xgb or lgtm, and fail for others. We could print a warning, in case the model type is different, but make sure that the code runs without errors. What i like about this solution is, the code and docs required is very low.
- Make a new wrapper class (as in your code example), which extends the ShapRFECV and overwrites the `_get_feature_shap_values_per_fold `. We could call the class `EarlyStoppingShapRFECV` or something similar. It is a pretty good solution, since it is clear to the users why they would use this class. However, we would write and maintain extra code and docs just this use case.

Both options seem good, i like the second option slightly more, because it is more usable. What do you think @operte  @timvink ?

Overall, this would be a crucial feature, because already multiple users asked about this.

**Comment by operte** [comment]
Indeed the second option is more explicit, but I would prefer the first. I think that we can make it clear in the explanation of the `early_stopping_rounds` argument that this only works with classifiers that allow early stopping. 

Let's see what @timvink thinks.

**Comment by operte** [comment]
Hmm now that I think about it, a problem with the solution I presented before is that we would have to add to `ShapRFECV` all other arguments that classifiers that use early stopping can accept (like `verbose` or `metric` etc.). 

An alternative would be to allow `ShapRFECV.fit` to pass additional arguments to the classifier. Then, inside `_get_feature_shap_values_per_fold`, we can adjust the fitting process depending if there is a `early_stopping_rounds` in the fit arguments or not. Something like this: 

```python
class ShapRFECVES(ShapRFECV):
    def _get_feature_shap_values_per_fold(self, X, y, clf, train_index, val_index, scorer, verbose=0, **kwargs):
        X_train, X_val = X.iloc[train_index, :], X.iloc[val_index, :]
        y_train, y_val = y.iloc[train_index], y.iloc[val_index]
        
        # Fit model with train folds
        if "early_stopping_rounds" in kwargs.keys():
            clf = clf.fit(
                X_train, y_train,
                eval_set=[(X_train, y_train), (X_val, y_val)],
                **kwargs
            )
        else:
            clf = clf.fit(X_train, y_train)
  
        # Score the model
        score_train = scorer(clf, X_train, y_train)
        score_val = scorer(clf, X_val, y_val)
  
        # Compute SHAP values
        shap_values = shap_calc(clf, X_val, verbose=verbose)
        return shap_values, score_train, score_val`

```

**Comment by Matgrb** [comment]
That is an interesting option. However,  typically the kwargs** provided by the user are passed directly to the shap explainer. If we start using the kwargs** in two places this may decrease the usability and consistency. that is why i would not combine it.

I think it is easier to either pass `early_stopping_rounds ` argument to init, or have a separate class.

**Comment by operte** [comment]
I'm not sure what you mean, but here the  kwargs would be set only on `.fit`. I see that `.fit` doesn't accept any other arguments, so I think it would be fine.

**Comment by Matgrb** [comment]
If you look at the fit() in `ShapModelInterpreter` you will see argument **shap_kwargs that is passed to the shap explainer. This way the user can modify how shap works. I was sure that the `ShapRFECV`and `SHAPImportanceResemblance` implement the same parameter, but i see now that it is not. That caused the confusion, sorry.

I think it would be quite handy, especially for the second one, since the user might want to modify how the explainer is applied. 

Having said that, we could consider where to go from here:
- We could allow the user to pass `**model_fit_kwargs` in `fit()` to enable the user to modify the fit parameters. This would solve the problem of early stopping, but decrease the consistency a bit.
- We could stick to the `**shap_kwargs` for all SHAP related features, for the consistency, and figure out a different way to pass model fit parameter.

**Comment by timvink** [comment]
Bit late to the party. I like the solution of allowing both `shap_kwargs` and `model_fit_kwargs` on `ShapRFECV.__init__ `, it will allow the user to more flexibility to make small changes.

Early stopping seems most useful for determining the number of rounds for LightGBM. Perhaps a better approach would be to use the existing gridsearch functionality (passing a `GridSearchCV` as a `clf`) to first determine the optimal number of rounds.

You could use ray's [TuneGridSearchCV](https://github.com/ray-project/tune-sklearn#tunegridsearchcv) that is a drop-in replacement, but adds an extra `early_stopping` parameter. So once #76 is solved, we might be able to get away with just updating the docs with an example for this issue.

**Comment by Matgrb** [comment]
I really like the solution - it requires the least code changes, and provides the most flexibility!

**Comment by operte** [comment]
I don't see how this solves the problem that for LGBM and XGB one has to pass a `valid_set` to the fit model. We would still need to have a GridSearchCV that allows for this, right?

**Comment by Matgrb** [comment]
If we implement the #76 we will allow for any SearchCV that is compatible with sklearn. 

Then the user can pass something like this as `clf`:
```
# Example parameters to tune from SGDClassifier
parameters = {
    'alpha': [1e-4, 1e-1, 1],
    'epsilon':[0.01, 0.1]
}

tune_search = TuneGridSearchCV(
    SGDClassifier(),
    parameters,
    early_stopping="MedianStoppingRule",
    max_iters=10
)
```

**Comment by Matgrb** [comment]
Hmm, but i just read more about the method that we refered to. The early stopping there is to stop hyperparameter search if it is not bringing gains. However, it still does not solve the problem of stopping fitting the XGBoost.

In the end, i think if we want to implement early stopping on the model side, we need to implement one of the following:
- wrapper class `EarlyStoppingShapRFECV`
- `early_stopping_rounds` to `ShapRFECV.__init__`
- Allow for `**fit_kwargs` while making sure that they are passed correctly to `clf.fit()`

I am most in favour of the first option, just to keep the API of the original `ShapRFECV` as clean as possible, what do you think?

**Comment by operte** [comment]
I think that even for #76, the problem remains that when one runs `ShapRFECV.fit`, this calls `ShapRFECV._get_feature_shap_values_per_fold` with train and valid sets as parameters, which is all good, but then inside this latter method, `clf.fit` is called without passing the validation set. So it appears to me that even to implement #76 we would need to change `ShapRFECV._get_feature_shap_values_per_fold`.

**Comment by operte** [comment]
I am growing fonder of the wrapper solution. Even if we allowed the user to pass `**fit_kwargs`, there would still need to be some tuning of `_get_feature_shap_values_per_fold` to enable, for example, using `eval_sample_weight` or `eval_init_score` from [LGBM's `.fit`](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.LGBMClassifier.html#lightgbm.LGBMClassifier.fit). 
It doesn't feel realistic for us to have to maintain that. 

I'll write a wrapper based on [my previous comment](https://github.com/ing-bank/probatus/issues/39#issuecomment-780526410). It implements simple early stopping, which is what we mostly use at our chapter. We can write a tutorial on this wrapper class and encourage the users to modify it for their particular needs.

**Comment by Matgrb** [comment]
Nice idea, let's start with simple wrapper that we can have as a separate tutorial notebook.

If in the future, the interest grows in this feature, we can add as part of probatus as well. Feel free to open the PR for that issue.

**Comment by Matgrb** [comment]
Since this is a specific use case, I would even suggest making a notebook and a separate tab in the Probatus documentation.

The tab would be How-to (according to [DIVIO)](https://docs.divio.com/en/latest/)
And there we can put such specific examples:
"How to apply ShapRFECV with LightGBM using Early stopping"

**Comment by operte** [comment]
Yes, I agree with that. But would you then have the wrapper only on the notebook or as a part of the package itself?

**Comment by Matgrb** [comment]
It would be better to have it as part of the package itself as something like `EarlyStoppingShapRFECV `. That would mean that implementation also requires adding a couple of unit tests, and good docstring. This way we can get this into a more maintainable state.

However, a notebook only is a good starting point as well,  and then opening another issue to merge it into a package. It is up to you how much spare time you have to implement it.
